### PR TITLE
storage: enable partial images by default

### PR DESCRIPTION
**Pull request by giuseppe**
by default enable pulling a partial image, it is still possible to disable the feature through the configuration file.

@rhatdan @mtrmac

### CODE REVIEW

**Review by mtrmac — COMMENTED** [review]
Code LGTM; but doing this is ~blocked on #1826 and c/image using that API, otherwise power loss can lead to broken stores.

---

As discussed in chat, having the check for `enable_partial_images` at the very first entry point (`chunked.GetDiffer`) would isolate more of the risk.

**Comment by mtrmac** [comment]
FWIW tests are failing

**Comment by giuseppe** [comment]
fixed now

File: pkg/chunked/storage_linux.go [hunk 1]
```diff
@@ -259,10 +263,6 @@ func GetDiffer(ctx context.Context, store storage.Store, blobDigest digest.Diges
 }
 
 func makeConvertFromRawDiffer(ctx context.Context, store storage.Store, blobDigest digest.Digest, blobSize int64, annotations map[string]string, iss ImageSourceSeekable, storeOpts *types.StoreOptions) (*chunkedDiffer, error) {
-	if !parseBooleanPullOption(storeOpts, "convert_images", false) {
```
Comment (mtrmac) [inline pkg/chunked/storage_linux.go:262]: Was this removal intentional?

Comment (giuseppe) [reply pkg/chunked/storage_linux.go]: no sorry, that was a mistake.  Re-added

**Review by mtrmac — COMMENTED** [review]
Code LGTM again… still #1826.

File: docs/containers-storage.conf.5.md [hunk 2]
```diff
@@ -96,7 +96,7 @@ storage of content and can allow the kernel to use less memory when running
 containers.
 
 containers/storage supports three keys
-  * enable_partial_images="true" | "false"
```
Comment (rhatdan) [inline docs/containers-storage.conf.5.md:99]: This looks wrong options is true or falsenot true or true.

File: docs/containers-storage.conf.5.md [hunk 3]
```diff
@@ -96,7 +96,7 @@ storage of content and can allow the kernel to use less memory when running
...
+  * enable_partial_images="true" | "true"
```
Comment (rhatdan) [inline docs/containers-storage.conf.5.md:99]: This looks wrong options is true or falsenot true or true.

Comment (giuseppe) [reply docs/containers-storage.conf.5.md]: fixed

**Comment by giuseppe** [comment]
> Code LGTM again… still #1826.

unblocked now

**Review by mtrmac — COMMENTED** [review]
/lgtm

**Comment by cgwalters** [comment]
Likely fallout in https://github.com/containers/bootc/issues/509#issuecomment-2156630865

**Comment by cgwalters** [comment]
Thanks for following up. Still to this day, we haven't really dug into zstd:chunked images and ostree because https://github.com/containers/bootc/issues/20 means it's basically not useful to create them right now - in theory it should be pullable (and most of the time was), but without the actual delta effect.

> But the total sequence of operations is not trivially clear from that report, and I’m naively thinking that if we broke podman pull && podman push, 

Right but remember "podman pull" is exercising the *combination* of c/image and c/storage which are wired together for zstd:chunked.

What ostree is doing is more like `skopeo copy docker://quay.io/some-zstd-chunked-image oci:ocidir` - and actually I guess not even that because that path goes through the whole "copy machinery" where c/image is driving things instead of being pull based. But it's at least a lot closer to "just fetch the blobs for an zstd:chunked images".

**Comment by cgwalters** [comment]
I guess we could try stress testing this path with just raw code talking to the proxy and calling GetBlob (https://github.com/containers/containers-image-proxy-rs/blob/main/examples/client.rs is a trivial client that doesn't involve anything ostree, etc.)

**Comment by mtrmac** [comment]
Hum… `skopeo copy docker://… non-c/storage:…` would not trigger the chunked path at all. That _is_ less likely to be exercised in existing tests, but it’s also simple enough in implementation that I don’t see anything that could obviously be wrong when creating +pushing, or consuming zstd:chunked layers on registries.
